crates/shared/src/token_list.rs

```diff
@@ -1,12 +1,45 @@
-use std::collections::HashMap;
+use std::{
+    collections::{HashMap, HashSet},
+    sync::{Arc, RwLock},
+    time::Duration,
+};
 
 use anyhow::Result;
 use ethcontract::H160;
-use reqwest::{Client, IntoUrl};
+use reqwest::Client;
 use serde::Deserialize;
 
+#[derive(Clone, Debug, Default)]
+pub struct TokenListConfiguration {
+    pub url: String,
+    pub chain_id: u64,
+    pub client: Client,
+    pub update_interval: Duration,
+}
+
+impl TokenListConfiguration {
+    async fn tokens(&self) -> Result<HashMap<H160, Token>> {
+        let model: TokenListModel = self
+            .client
+            .get(self.url.clone())
+            .send()
+            .await?
+            .json()
+            .await?;
+        Ok(Self::from_tokens(model.tokens, self.chain_id))
+    }
+
+    fn from_tokens(tokens: Vec<TokenModel>, chain_id: u64) -> HashMap<H160, Token> {
+        tokens
+            .into_iter()
+            .filter(|token| token.chain_id == chain_id)
+            .map(|token| (token.token.address, token.token))
+            .collect()
+    }
+}
+#[derive(Clone, Debug, Default)]
 pub struct TokenList {
```

Comment:
We should document that this contains a shared and auto updating token list since it's not clear from the type name.

Reply:
I've renamed the struct to `AutoUpdatingTokenList`. I think I also used this kind of naming for other functionalities that had background tasks updating it's state so feels right.